### ISSUE TITLE
Using Valhalla as a Library May Miss Proper Includes

### ISSUE THREAD

**Issue by kevinkreiser** [valhalla/valhalla#3747]
over in https://github.com/valhalla/valhalla/pull/2192 we tried to address this issue but landed on a different tactic. the essential problem is that we are vendoring header only libraries at build time but when someone downstream links us and builds against our headers they dont get the vendored headers as well. we need to not only vendor the headers when we install our packages we should also change the way we  include them to be the full path `<valhalla/third_party/...>` so that we make sure downstream users are using the same header they built the library with.

related also to #3734

**Comment by nilsnolde** [comment]
I also wonder if it wouldn’t make sense to not vendor rapidjson anymore and instead rely on package managers? In the event of releasing a Valhalla package for a distribution, I guess it’d be a problem if people also have it installed via their package manager (as rapidjson is public API). I didn’t review yet existing versions for apt-get/dnf/.. and the supported distros. 

Smth we could of course also do once we get there. Just popped into my head when reading this.

**Comment by kevinkreiser** [comment]
Vendeoring is a better solution in my opinion as it pins the version. We just have to actually do it correctly :)

**Comment by nilsnolde** [comment]
It’s def nicer for us :) 

I think from a package manager’s perspective it’s not so great when the package is also available as system package (as rapidjson is by now for almost all distros), in my head public deps would create conflicts if they’re both vendored and system-wide (eg as dependency of another system app/lib). 

But yeah, I’ll get back to it once I aspire to package Valhalla for any distro (mid-term I’m considering pacman) :) or there’ll be a tiny patch file.. never mind me😅

**Comment by kevinkreiser** [comment]
> in my head public deps would create conflicts if they’re both vendored and system-wide (eg as dependency of another system app/lib)

yep thats my point, if we vendored them correctly they couldnt conflict, specifically by including them from the thirdparty directory and not elsewhere. eg. `#include <valhalla/third_party/rapidjson/include/rapidjson.h>` then you cannot have a conflict unless some system include or other system library exposes them in their headers and we get them inadvertently. this does bring up another long term thing we should be doing better with which is not leaking thirdparty includes in our headers which goes along the lines of reducing the scope of our headers to only expose the bits of the library that makes sense for people who want to use it as such and not just throwing every function into the header because its convenient. honestly we could have used a lot more discipline with this over the years so the effort to clean it up now is herculean haha
